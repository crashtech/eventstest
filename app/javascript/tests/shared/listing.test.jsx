import React from 'react'
import renderer from 'react-test-renderer'

import Listing from '../../shared/listing'

const itemRenderer = (item) => item
const dateRenderer = ({ text }) => text

test('Simple listing result', () => {
  const data = ['Item 1', 'Item 2', 'Item 3']
  const fetcher = () => Promise.resolve({ data: [], headers: { 'x-chunks': 2 }, update: () => {} })
  const component = renderer.create(
    <Listing
      items={data}
      fetcher={fetcher}
      renderer={itemRenderer}
    />,
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('Can load new records', () => {
  let data = []
  let promise = null
  const fetcher = () => promise = Promise.resolve({
    data: ['Item 1', 'Item 2'],
    headers: { 'x-chunks': 2 },
    update: (items) => data = items,
  })

  const component = renderer.create(<Listing items={data} fetcher={fetcher} renderer={itemRenderer} />)
  expect(component.toJSON()).toMatchSnapshot()

  promise.then(() => {
    renderer.update(<Listing items={data} fetcher={fetcher} renderer={itemRenderer} />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})

test('Can have a fallback', () => {
  const data = []
  const fetcher = () => Promise.resolve({ data: [], headers: {}, update: () => {} })
  const component = renderer.create(
    <Listing
      items={data}
      fetcher={fetcher}
      fallback='No records'
      renderer={itemRenderer}
    />,
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('Can be divided by day', () => {
  const data = [
    { text: 'Item 1', date: '2020-01-01' },
    { text: 'Item 2', date: '2020-01-01' },
    { text: 'Item 3', date: '2020-01-02' },
  ]

  const fetcher = () => Promise.resolve({ data: [], headers: {}, update: () => {} })
  const component = renderer.create(
    <Listing
      items={data}
      fetcher={fetcher}
      renderer={dateRenderer}
      dividedBy="day"
      dividedProp="date"
    />,
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('Can be divided by month', () => {
  const data = [
    { text: 'Item 1', date: '2020-01-01' },
    { text: 'Item 2', date: '2020-02-01' },
    { text: 'Item 3', date: '2020-03-01' },
  ]

  const fetcher = () => Promise.resolve({ data: [], headers: {}, update: () => {} })
  const component = renderer.create(
    <Listing
      items={data}
      fetcher={fetcher}
      renderer={dateRenderer}
      dividedBy="month"
      dividedProp="date"
    />,
  )

  expect(component.toJSON()).toMatchSnapshot()
})
