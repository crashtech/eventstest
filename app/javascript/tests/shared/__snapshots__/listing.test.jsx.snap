// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can be divided by day 1`] = `
<div
  className="ui basic segment listing"
>
  <div
    className="ui divided list"
    role="list"
  >
    <div
      className="item scale divider visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      <div
        className="ui horizontal divider"
      >
        January 1, 2020
      </div>
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 1
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 2
    </div>
    <div
      className="item scale divider visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      <div
        className="ui horizontal divider"
      >
        January 2, 2020
      </div>
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 3
    </div>
  </div>
  <div />
</div>
`;

exports[`Can be divided by month 1`] = `
<div
  className="ui basic segment listing"
>
  <div
    className="ui divided list"
    role="list"
  >
    <div
      className="item scale divider visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      <div
        className="ui horizontal divider"
      >
        January 2020
      </div>
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 1
    </div>
    <div
      className="item scale divider visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      <div
        className="ui horizontal divider"
      >
        February 2020
      </div>
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 2
    </div>
    <div
      className="item scale divider visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      <div
        className="ui horizontal divider"
      >
        March 2020
      </div>
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 3
    </div>
  </div>
  <div />
</div>
`;

exports[`Can have a fallback 1`] = `
<div
  className="ui basic loading segment empty listing"
>
  <div
    className="ui list"
    role="list"
  >
    <div
      className="item scale fallback animating out visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": "300ms",
        }
      }
    >
      No records
    </div>
  </div>
  <div />
</div>
`;

exports[`Can load new records 1`] = `
<div
  className="ui basic loading segment empty listing"
>
  <div
    className="ui list"
    role="list"
  />
  <div />
</div>
`;

exports[`Simple listing result 1`] = `
<div
  className="ui basic segment listing"
>
  <div
    className="ui list"
    role="list"
  >
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 1
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 2
    </div>
    <div
      className="item scale visible transition"
      onClick={[Function]}
      role="listitem"
      style={
        Object {
          "animationDuration": undefined,
        }
      }
    >
      Item 3
    </div>
  </div>
  <div />
</div>
`;
